src/linear_algebra/nonsingular_inverse.lean

```diff
@@ -0,0 +1,300 @@
+for `matrix` so that can be used for the pseudoinverse instead.
+
+The definition of inverse used in this file is the one given by Cramer's rule.
+The vectors returned by Cramer's rule are given by the linear map `cramer`.
+Using Cramer's rule, we can compute for each matrix `A` the matrix `adjugate A`,
+which we then prove behaves like `det A • A⁻¹`. Finally, we show that dividing
+the adjugate by `det A` (if possible), giving a matrix `nonsing_inv A`, will
+result in a multiplicative inverse to `A`.
+
+## References
+
+  * https://en.wikipedia.org/wiki/Cramer's_rule#Finding_inverse_matrix
+
+## Tags
+
+matrix inverse, cramer, cramer's rule, adjugate
+-/
+
+namespace matrix
+universes u v
+variables {n : Type u} [fintype n] [decidable_eq n] {α : Type v}
+open_locale matrix
+open equiv equiv.perm finset
+
+section replace
+
+/-- Replace the `i`th column of matrix `A` with the values in `b`. -/
+def replace_column (A : matrix n n α) (i : n) (b : n → α) : matrix n n α :=
+function.update A i b
+
+/-- Replace the `i`th row of matrix `A` with the values in `b`. -/
+def replace_row (A : matrix n n α) (j : n) (b : n → α) : matrix n n α :=
+λ i, function.update (A i) j (b i)
+
+variables {A : matrix n n α} {i j : n} {b : n → α}
+
+@[simp] lemma replace_column_self : replace_column A i b i = b := function.update_same i b A
+@[simp] lemma replace_row_self : replace_row A j b i j = b i := function.update_same j (b i) (A i)
+
+@[simp] lemma replace_column_eq {i' : n} (h : i = i') : replace_column A i b i' = b :=
+by {rw [h], apply replace_column_self}
+@[simp] lemma replace_row_eq {j' : n} (h : j = j') : replace_row A j b i j' = b i :=
+by {rw [h], apply replace_row_self}
+
+@[simp] lemma replace_column_ne {i' : n} (i_ne : i' ≠ i) : replace_column A i b i' = A i' :=
+function.update_noteq i_ne b A
+@[simp] lemma replace_row_ne {j' : n} (j_ne : j' ≠ j) : replace_row A j b i j' = A i j' :=
+function.update_noteq j_ne (b i) (A i)
+
+lemma replace_column_val {i' : n} : replace_column A i b i' j = if i' = i then b j else A i' j :=
+begin
+  by_cases i' = i,
+  { rw [h, replace_column_self, if_pos rfl] },
+  { rw [replace_column_ne h, if_neg h] }
+end
+lemma replace_row_val {j' : n} : replace_row A j b i j' = if j' = j then b i else A i j' :=
+begin
+  by_cases j' = j,
+  { rw [h, replace_row_self, if_pos rfl] },
+  { rw [replace_row_ne h, if_neg h] }
+end
+
+lemma replace_column_transpose : replace_column Aᵀ i b = (replace_row A i b)ᵀ :=
+begin
+  ext i' j,
+  rw [transpose_val, replace_column_val, replace_row_val],
+  refl
+end
+end replace
+
+section cramer
+/-!
+  ### `cramer` section
+
+  Introduce the linear map `cramer` with values defined by `cramer_map`.
+  After defining `cramer_map` and showing it is linear,
+  we will restrict our proofs to using `cramer`.
+-/
+variables [comm_ring α] (A : matrix n n α) (b : n → α)
+
+/--
+  The `cramer_map` sends a matrix `A` and vector `b` to the vector `x`,
+  such that `A ⬝ x = b`.
+-/
```

Comment:
I don't see how the docstring relates to the definition (in general, if `A` is not onto, there is no `x` with `Ax=b`...)